### DISCUSSION TITLE
Parent-child relationships for Competency Management

### DISCUSSION THREAD

**Discussion by sagarihanda** [Sunbird-Knowlg/Community#45] [Documentation Issues]
### Project Context
UP State Medical Faculty (SMF) is building a platform to manage human resources for allied healthcare professionals (nurses, paramedical staff, lab technicians etc.) in the state. The Government is looking to build the platform on Sunbird building blocks to adhere to ABDM's (https://abdm.gov.in/abdm) principles. One of the key modules being built is to enable competency based assessments of allied professionals, and to enable mid-career learning courses via an LMS. Six Sunbird blocks (RC, Obsrv, Knowlg, inQuiry, CoKreat and Ed) are being leveraged to develop the platform with the user data being housed in Sunbird RC.

### Requirement
We are looking to implement a competency tree as per [FRAC](https://clopinitiative.org/sites/default/files/webinar/Session%201_%20Webinar_The%20Framework%20of%20Roles%2C%20Activities%20and%20Competencies%20%28FRAC%29-speaker%20notes_AR_NJ.pdf) framework of DOPT (Department of Personnel and Training) which include four levels - Functions > Roles > Activities > Competencies. Additionally, each Competency has multiple levels (3 to 5) in progressing order of difficulty / complexity. See visual representation below.

![Competency Management UP@2x (1)](https://user-images.githubusercontent.com/73831409/182788557-1b7e1d66-399f-4e55-8584-02532516e873.png)

### Question
We understand that it is possible to create custom _Categories_ and _Objects_ to model parent-child relationships in the competency framework. But, are unable to find documentation or an example showcasing how this can be done on the Knowlg repo. Can the community share an example of how this can be done? 

@ChakshuGautam @rahul101001000 @reshmi-nair @rhwarrier @pkranga @Mahesh-Maney

**Comment by rhwarrier** [comment]
@vinukumar-vs  @maheshkumargangula Please share the necessary inputs. Thanks

**Comment by vinukumar-vs** [comment]
Please find the below reference links to know more details about Framework.
https://knowlg.sunbird.org/learn/product-and-developer-guide/taxonomy-and-tagging/framework-service (Go through the attached video)
https://diksha.gov.in/help/faqs/creator/taxonomy-and-framework.html 

@kameshbhr Can you please share the Diksha framework details if it is documented?

@vrayulu, @surendrasinghs, @pallakartheekreddy

**Reply by ChakshuGautam** [reply]
@vinukumar-vs @rhwarrier @vrayulu, As discussed on the call yesterday, it would be great if we have a couple of curl requests showing this as well. We have gone through the docs but would need an opinionated way of making this work.

**Reply by Jayaprakash8887** [reply]
https://knowlg.sunbird.org/learn/product-and-developer-guide/taxonomy-and-tagging/framework-service/faqs
Please refer to the above link.

**Comment by ChakshuGautam** [comment]
Based on the call today to figure out the challenges of implementing FRAC through Sunbird `framework`, sharing the meeting notes here. @vinukumar-vs @pkranga @mathewjpallan

### Challenges
1. *Data Storage and APIs*
	- When the framework is read a single huge list is returned in some cases 10MB (Works well with a limited data set).
	- The node at level 1 has all the associations till the nth level returned even for the limited set of fields.
2. *Search*: Limited search capabilities for terms as they are not indexed in ES.
3. *Version Control*: No management is possible as there is no concept of version control.
4. *UI*:  for workflow for approval.

Please see the [Reference Doc](https://docs.google.com/document/d/1fVPNCFkP4evqR0BFHcf6vpM3WWsYbX7U8LjqUNAQ7zg/edit?usp=sharing) for more details.

Based on the above challenges 1, 2, and 3, @vinukumar-vs we should once check with the DC if this is something that should become part of the Framework Service since without these it is becoming difficult to model FRAC. 

UI can sit outside the Sunbird for now.

**Reply by sukhpreetssekhon** [reply]
Thanks for the summary @ChakshuGautam 

@mathewjpallan - Please can you initiate the RFC for points 1-3. 
@Krishnaj20 - Can you review and drive the closure for the RFC with @vinukumar-vs 
@vinukumar-vs - Can you layout next steps if we are to close next steps with DC within a week
@pkranga - Can you please share any more details of the existing implementation, that is publicly shareable, regarding what has been implemented so far in FRAC
@coolbung - Do share the wireframes for point 4 based on the use cases you are planning to implement.

cc - @surendrasinghs @vrayulu @rahul101001000 @pramodkvarma

**Reply by vrayulu** [reply]
@sukhpreetsamagra @ChakshuGautam 

The challenges listed above are not accurate:

1. Data Storage and APIs

- Framework read does not return all the list always. There is an option to specify the categories that are to be returned: <host_name>/api/framework/v1/read/<framework_id>?categories=board,gradeLevel,medium,class,subject. Using this, we can get the subset of the framework as needed.
- Did not understand what this means "The node at level 1 has all the associations till the nth level returned even for the limited set of fields." Please share more details.

2. Terms are indexed into ES and can be searched using composite search API.

3. There is no versioning but publish workflow is available. Very small change to bump up the version number on every publish. If the capability to fetch previous versions is needed, that can also be enabled - similar thing is already designed for Content objects.

4. Yes, UI is not available. But why does it need sit outside Sunbird. Similar to how players and editors are built as components that can be embedded into other portals, we should build Taxonomy management UI components and have them as part of Knowlg BB.

**Reply by sukhpreetssekhon** [reply]
Thanks for the update @vrayulu.

@mathewjpallan - Does this provide better clarity? Any further clarifications?

On the point regarding UI, completely agree @vrayulu. We should have UI as part of Sunbird. In fact @coolbung is building a very similar workflow based use case on Framework and plan to contribute the UI back to Sunbird as well which we can leverage here. @coolbung - do add any wireframes that you may have already.

**Reply by coolbung** [reply]
Sharing the draft of the wireframes
https://docs.google.com/presentation/d/1ThC3pd8ZcQ2ajzOqFk8JGWD8eJ4KdKcwbY6619Zhi9U/edit

@vrayulu -- the framework creation expects channel as a mandatory input. Is that intentional ? Since channels can be independently updated to set the default and additional frameworks, can this be removed ?

**Reply by vrayulu** [reply]
@coolbung - I think it is there to figure out who owns the framework... instead of channel id as input, it could be derived from API token. A similar approach is used for private content read API, we should adopt the same strategy here also. 

@vinukumar-vs - please share how private content is handled by knowlg currently.

**Reply by Jayaprakash8887** [reply]
@coolbung For a private content use-case, where only frameworks related to a root Org/tenant is to be displayed for their content creation/consumption, then channel can be used to filter out frameworks to be displayed for choosing.
